File: src/gen-ai.ts
import { GoogleGenAI, ThinkingLevel, GenerateContentConfig, Content, GenerateContentResponse, Part, MediaResolution } from '@google/genai';
import { geminiResponseSchema } from './gemini-response-schema';
import { EventData, EventDataSchema } from './types';

export async function extractEventData(imageBase64: string, caption: string): Promise<EventData> {
  const genAI: GoogleGenAI = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  const config: GenerateContentConfig = {
    thinkingConfig: {
      thinkingLevel: ThinkingLevel.MINIMAL,
    },
    responseMimeType: 'application/json',
    responseSchema: geminiResponseSchema,
    mediaResolution: MediaResolution.MEDIA_RESOLUTION_MEDIUM,
  };
  const model: string = 'gemini-3.1-flash-lite-preview';
  let parts: Part[] = [
    {
      text: `This is an IG post promoting an event. Extract structured event details from the image and caption below. Use UK local time for all datetimes. Omit fields not explicitly stated.\n\n--- IG CAPTION ---\n${caption}\n--- END IG CAPTION ---`,
    },
  ]

  if (imageBase64) {
    parts.unshift({ inlineData: { mimeType: 'image/jpeg', data: imageBase64 } });
  }

  const contents: Content[] = [{
        role: 'user',
        parts,
    }];

  const response: GenerateContentResponse = await genAI.models.generateContent({
    config,
    model,
    contents,
  });

  return EventDataSchema.parse(JSON.parse(response.text ?? ''));
}

File: src/gen_ai.py
import base64
import json
import os

import requests
from google.genai import Client
from google.genai.types import Blob, GenerateContentConfig, Part, ThinkingConfig

from src.gemini_response_schema import gemini_response_schema


def extract_event_data(image_url: str, caption: str) -> dict:
    client = Client(api_key=os.environ["GEMINI_API_KEY"])

    image_data = base64.b64encode(requests.get(image_url, timeout=10).content).decode("utf-8")

    prompt = (
        "This is an Instagram post promoting an event. "
        "Extract structured event details from the image and caption below. "
        "Use UK local time for all datetimes. Omit fields not explicitly stated.\n\n"
        f"--- INSTAGRAM CAPTION ---\n{caption}\n--- END INSTAGRAM CAPTION ---"
    )

    response = client.models.generate_content(
        model="gemini-3.1-flash-lite-preview",
        contents=[
            Part(inline_data=Blob(mime_type="image/jpeg", data=image_data)),
            prompt,
        ],
        config=GenerateContentConfig(
            response_mime_type="application/json",
            response_schema=gemini_response_schema,
            thinking_config=ThinkingConfig(thinking_budget=0),
        ),
    )

    return json.loads(response.text)
